# Aldagaiak

Pythonen, beste rogramazio lengoaietan bezala, balio berriak sortuko dituzten espresioak existitzen dira:

In [1]:
365/7

52.142857142857146

Defektuz, balio hauek *galdu* egiten dira. Balio hauek aldabaietan *"gorde"* ditzakegu, aurrerago erabili ahal izateko:

In [2]:
aste = 365/7
print("Urteak",aste,"ditu")

Urteak 52.142857142857146 ditu


Balioak gordetzeko, aldagaiak eta esleipen sententziak erabiltzen dira. Esleipen sententzi batek ondoko pausuak ematen ditu:

  1. Eskuinaldeko espresioaren balioa kalkulatu.
  1. Ezkerreko aldagaiari esleitu balio hori.

Tira, hau ez da guztiz zuzena...
![scream](../img/emoji/36x36/1f631.png)

## Aldagaiak erreferentziak dira

Aldagaiak **erreferentzia edo etiketa soilak** dira. Aldagaietan gordetzen den informazioa ez da esleitzen dugun balioa, baizik eta **balio horretara iristeko behar den informazioa** (objektu horren helbidea).

`aste = 365/7` sententzia exekutatzean, ondoko pausuak ematen dira:

  - Eskuinaldeko espresioa ebaluatzen da.
    - `365` eta `7` balioak izango dituzten bi objektuak (zenbaki osoak) sortzen dira.
    - Zatiketa eragiteta ebaluatzean, `52.142857142857146` balioa izango duen objektua (zenbaki erreala) sortzen da.
  - `aste` aldagaiari `52.142857142857146` objektuaren erreferentzia esleitzen zaio.
  - Hemendik aurrera, `aste` identifikadorea erabiltzen dugun guztietan, `52.142857142857146` balioa duen objektuaz ari gara hitzegiten.
    
![VariablesAndReferences1](../img/VariablesAndReferences1.png)


**Aldaezinak** diren informazio edo balioekin ari garenean, berdin digu aldagaiak erreferentziak direla pentsatzea edo ta aldagaiek balioak gordetzen dituztela pentsatzea. Demagun `a` aldagai bati `1234` balioa esleitzen diogula,

`a = 1234`

eta ondoren `b` aldagai bati `a`-ren balioa esleitzen diogula,

`b = a`

Ondoko irudian sententziak exekutatu ondorengo bi egoera agertzen dira, bi ikuspuntuen arabera (ezkerrean aldagaiek balioak gordetzen dituzte eta eskuinean balioen erreferentziak).

![VariablesAndReferences2](../img/VariablesAndReferences2.png)


Aldagaietan balioak gordeko balira (ezkerraldea), bigarren sententziak `1234`-ren kopia bat gordeko luke `b`-ren barnean. Aldagaiak erreferentziak balira ordea (eskuinaldea), bigarren sententziak ez luke objektu berririk sortuko (ez dago inongo eragiketarik), eta beraz `a` eta `b` erreferentziek objektu berdin bat erreferentziatuko lukete (bi erreferentziek helbide berdina dute). Ezberdinak badira ere, bi egoerak baliokideak dira eta programazioaren ikuspuntutik berdin dio modu batera edo bestera pentsatzeak.

**Aldakorrak** diren informazioetan, ordea, bi egoerak guztiz ezberdinak dira. Demagun zenbaki osoen informazioa ez dela soilik balio oso bat, demagun kolorea ere badutela, eta kolorez aldatu daitezkeela (`1234` zenbaki oso literala adieraztean, pythonek defektuzko kolore beltza erabili duela pentsa genezake). Demagun ondoko sententziak `a.kolorea_aldatu("orange")` a objektuaren kolorea aldatzen duela. Ondokoa litzateke emaitza:

![VariablesAndReferences3](../img/VariablesAndReferences3.png)


Argi ikusten denez, aldagaiek balioak gordeko balituzte, `a`-ren kolorea aldatzeak ez luke `b`-ren kolorea aldatuko. Erreferentziak direnez, ordea, `a` eta `b` erreferentziek objektu berdin bat erreferentziatzen dute, eta beraz `a`-ren kolorea aldatzean `b`-ren kolorea aldatu dugu ere.

<table border="0" width="100%" style="margin: 0px;">
<tr>  
  <td style="text-align:left"><a href="Balio literalak eta datu motak.ipynb">&lt; &lt; Balio literalak eta datu motak &lt; &lt;</a></td>
  <td style="text-align:right"><a href="Aldagai izenak.ipynb">&gt; &gt; Aldagai izenak &gt; &gt;</a></td>
</tr>
</table>